### ISSUE TITLE
Hide `window` using an AST transform

### ISSUE THREAD

**Issue by kevinbarabash** [Khan/live-editor#547]
There are certain properties on window that we'd like to hide from users such as `console`, `location`, `document` (at least for processing-js), etc.  Using regexes to ban these terms prevents their use in comments and strings.  Searching for identifier nodes in the AST with these names is a little better but it's still not perfect because it might make sense to have a `location` property on an object.  The proposed solution avoids these inaccuracies by replacing the `window` object with our own object that exposes only those functions/properties we want to expose.

Basic idea: transform
```
function Foo() {
    this.bar = "hello, world";
}
```
into
```
function Foo() {
    var that =  this === window ? myWindow : this;
    that.bar = "hello, world";
}
```

In order to deal with someone redefining `that` we should rewrite all user variable references to be properties on a local `scope` variable, e.g. transform
```
function Foo() {
    var x = 5;
    var y = 10;
}
```
into
```
function Foo() {
    var scope = {};
    scope.x = 5;
    scope.y = 10;
}
```

**Comment by DozenalFTW** [comment]
This seems too complicated... why not just redefine your `window` object to null the parameters you don't need?

**Comment by kevinbarabash** [comment]
In the case of `window.location` it will cause the page to reload.  There might be other properties we don't want users to access, but they're read-only.

**Comment by DozenalFTW** [comment]
Possibly, you could change the getter and setter of `location`?
You could also define a ridiculous random identifier, like `__thisObj__9eH32K40A`, and generate a different one if that identifier is already taken. No need for `scope`.

**Comment by kevinbarabash** [comment]
> You could also define a ridiculous random identifier, like __thisObj__9eH32K40A

That works too.

**Comment by Potato42** [comment]
This won't work.  As long as you can get at some form of eval, you can bypass the AST completely.  Maybe if you completely restricted access to all forms of eval (there's a CSP option for that), you might have a chance at preventing users from getting at anything they want in the window object, but otherwise it will simply be an annoyance, and eval is used for plenty benign programs I've seen on KA.  Not only that, but I imagine `document.createElement("script")` would be a form of eval not blocked by the CSP that, outside of the pjs environment, wouldn't make sense to block because document.createElement (and equivalents in jQuery) are an important part of js outside the pjs environment.  Also, there are ways to get at things via processing.js builtins.  For example, `createFont("some-font").context2d.canvas.ownerDocument` gets you `document`.  Perhaps you could use the ASTs to check the result of each `.` or `[]` operation, but that would be a little ridiculous.  Scope (if I'm understanding what you mean correctly) is probably a bad idea because you would need to deal with out-of-scope variables, although I don't see anything wrong with the alternative of the random identifier.  Repeating @DozenalFTW's question, what is wrong with `console`?  Obviously it's not supported in all js environments, but so are a lot of things.

**Comment by kevinbarabash** [comment]
@Potato42 we don't allow `eval`.  Can you post a program that uses `eval` on KA?  I didn't realize that `createFont` return on object that had access to the canvas.  I wouldn't want to check the result of each `.` or `[]` operation.

I have some code that keeps a stack of all of the `scope` objects and whenever it encounters a variable reference it looks in each `scope` until it finds one that contains that variable, so really they'd be named `scope0`, `scope1`, etc.

I'm not sure why we block `console` since we alias `debug` to `console.log`.  I'll try to find out.

**Comment by Potato42** [comment]
> We don't allow eval

[Nonsense, only jshint "errors" keep you from using eval](https://www.khanacademy.org/computer-programming/eval/4999012598284288).  But I said all forms of eval.  This includes the `Function` constructor function (which can be accessed very easily from any function's constructor property - in fact it is a pseudo-standard on KA programs to use `Object.constructor` to access it), and as I mentioned, `document.createElement`.

`eval` is certainly not blocked on the webpage environment, where you don't even need a jshint comment to enable it.

**Comment by kevinbarabash** [comment]
All of these are amazing.  I especially like:
```
// jshint evil:true
```
I'm curious to see how many programs are using these work arounds.  @Potato42 are people escaping the sandbox for legitmate reasons?

**Comment by DozenalFTW** [comment]
@kevinbarabash 

> @Potato42 are people escaping the sandbox for legitmate reasons?

Yes, they are. I needed the `Date` object for [this](http://www.khanacademy.org/cs/if-you-want-to-chat/5992002041675776) because I wanted the UTC time, which `hour()` doesn't give you.

Let's not ban _all_ `jshint` tricks. I like disabling the warnings against `if (condition) statement;`, and when I know what I'm doing I don't need the `loopfunc` warning.

**Comment by kevinbarabash** [comment]
`Date` is totally legit.  It should be pretty easy to add, I just haven't got around to it.  Pull requests are welcome.  See https://github.com/Khan/live-editor/issues/520 for details.

As for `jshint`, I'm sure we can come up with a way expose certain `jshint` settings.

**Comment by Potato42** [comment]
"Legitimate reasons" is a little vague I guess, but I would say that most of the "legitimate" uses I've seen are in the category of getting around limitations and bugs in the pjs environment.
 * Getting accessed to external, but useful things.  The window object has a bunch of useful functions and data.  See @DozenalFTW's reason and consider along with it things like `window.atob`, `window.navigator.language/.userAgent`.
 * Working around bugs like [this one](https://www.khanacademy.org/computer-programming/ast-nonsense-safe-new/5440494445920256) does (although that bug is now fixed :grinning:).  I imagine messing with the ASTs too much to prevent users from using workarounds may cause more bugs while still failing to completely prevent all workarounds.
 * There are other similar things which are maybe less "legitimate", like loading a font from google-fonts in the PJS environment or otherwise loading things that pass the CSP but are not really in the intended scope of the PJS environment.

Eval itself has some legitimate uses.  It is often annoying to write an entire parser for some simple task, and although I don't know remember where I recall seeing it used for an educational-type program that had to do with bitwise operators a little while back.

**Comment by DozenalFTW** [comment]
Yeah, and what about Ben Burrill's `export_module` or that one mechanism that lets you have a font fallback system?

**Comment by DozenalFTW** [comment]
You see, I don't want to change every single variable into an object property. That would have loads of consequences:
```javascript
function () {
    __defineGetter__ = null;
    __defineSetter__ = null;
    prototype = null;
    constructor = null;
}
```
Just to name a few. I just want a window object with a random identifier _that is not inside the program_. So if we randomly generate `thisObj1817092` and that just so happens to appear, we'd regenerate a new identifier, possibly of longer length.

**Comment by kevinbarabash** [comment]
Good point about:
```
function () {
    __defineGetter__ = null;
    __defineSetter__ = null;
    prototype = null;
    constructor = null;
}
```
> So if we randomly generate thisObj1817092 and that just so happens to appear, we'd regenerate a new identifier, possibly of longer length.

That makes sense.  We should know all variable names before hand so coming up with an identifier that doesn't conflict is totally doable.

**Comment by kevinbarabash** [comment]
> I imagine messing with the ASTs too much to prevent users from using workarounds may cause more bugs while still failing to completely prevent all workarounds.

That's a really good point.  I think in general though that there are more issues from live updating not working quite right (or not quite as expected).

> Getting accessed to external, but useful things. The window object has a bunch of useful functions and data. See @DozenalFTW's reason and consider along with it things like window.atob, window.navigator.language/.userAgent, etc.

Yup... and `encodeURIComponent`, `JSON.stringify`, etc.

> Yeah, and what about Ben Burrill's export_module or that one mechanism that lets you have a font fallback system?

I appreciate the ingenuity in `export_module`, but I worry about people modifying dependencies which end up break other peoples' scratchpads.  I would like have built-in support for modules using CommonJS `require` and/or ES6 `import` syntax.  Ideally though there'd be a way to specify a version.  Before that though we'd want to expose a way to do versioning.

**Comment by DozenalFTW** [comment]
>> So if we randomly generate thisObj1817092 and that just so happens to appear, we'd regenerate a new identifier, possibly of longer length.

> That makes sense. We should know all variable names before hand so coming up with an identifier that doesn't conflict is totally doable.

```javascript
var userCode = getUserCodeHoweverYouWant();
var thisName = generateNewThisObject();
if (userCode.indexOf(thisName) !== -1) {
    lengthOfIdentifier++; // prevents SOMEHOW triggering an infinite loop!!
    fail();
} else {
    pass();
}
```

Who cares if it's in a string? It's so hard to hit something like this anyway.
